cudapoa/src/application_parameters.cpp

```diff
@@ -0,0 +1,226 @@
+/*
+* Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+*
+* NVIDIA CORPORATION and its licensors retain all intellectual property
+* and proprietary rights in and to this software, related documentation
+* and any modifications thereto.  Any use, reproduction, disclosure or
+* distribution of this software and related documentation without an express
+* license agreement from NVIDIA CORPORATION is strictly prohibited.
+*/
+
+#include "application_parameters.hpp"
+#include <fstream>
+#include <getopt.h>
+#include <iostream>
+#include <string>
+
+#include <claraparabricks/genomeworks/io/fasta_parser.hpp>
+#include <claraparabricks/genomeworks/utils/signed_integer_utils.hpp>
+#include <claraparabricks/genomeworks/version.hpp>
+
+namespace claraparabricks
+{
+
+namespace genomeworks
+{
+
+namespace cudapoa
+{
+
+ApplicationParameters::ApplicationParameters(int argc, char* argv[])
+{
+    struct option options[] = {
+        {"input", required_argument, 0, 'i'},
+        {"result", required_argument, 0, 'r'},
+        {"full-alignment", no_argument, 0, 'f'},
+        {"band-width", required_argument, 0, 'b'},
+        {"dot", required_argument, 0, 'd'},
+        {"max-groups", required_argument, 0, 'M'},
+        {"gpu-mem-alloc", required_argument, 0, 'R'},
+        {"match", required_argument, 0, 'm'},
+        {"mismatch", required_argument, 0, 'n'},
+        {"gap", required_argument, 0, 'g'},
+        {"version", no_argument, 0, 'v'},
+        {"help", no_argument, 0, 'h'},
+    };
+
+    std::string optstring = "i:r:fb:d:M:R:m:n:g:vh";
+
+    int32_t argument = 0;
+    while ((argument = getopt_long(argc, argv, optstring.c_str(), options, nullptr)) != -1)
+    {
+        switch (argument)
+        {
+        case 'i':
+            input_paths.push_back(std::string(optarg));
+            break;
+        case 'r':
+            result = std::stoi(optarg);
+            break;
+        case 'f':
+            banded = false;
+            break;
+        case 'b':
+            band_width = std::stoi(optarg);
+            break;
+        case 'd':
+            graph_output_path = std::string(optarg);
+            break;
+        case 'M':
+            max_groups = std::stoi(optarg);
+            break;
+        case 'R':
+            gpu_mem_allocation = std::stod(optarg);
+            break;
+        case 'm':
+            match_score = std::stoi(optarg);
+            break;
+        case 'g':
+            gap_score = std::stoi(optarg);
+            break;
+        case 'n':
+            mismatch_score = std::stoi(optarg);
+            break;
+        case 'v':
+            print_version();
+        case 'h':
+            help(0);
+        default:
+            exit(1);
+        }
+    }
+
+    if (gpu_mem_allocation <= 0 || gpu_mem_allocation > 1.0)
+    {
+        std::cerr << "gpu-mem-alloc should be greater than 0 and less than or equal to 1.0" << std::endl;
```

Comment:
looking through this again - let's throw a `runtime_error`exception in these cases?

Reply:
Changed for nearly everything except the input validation as it also prints the help message when 0 or invalid input is provided (useful to lookup usage).

Reply:
Perfect, thanks @atadkase !